### PR TITLE
[ddp] add enabled kwarg to no_sync

### LINKED ISSUE #143721: add "enabled=True" to DistributedDataParallel.no_sync()
### 🚀 The feature, motivation and pitch

Training a model with DDP and gradient accumulation is quite common.
To avoid unnecessary sync, the no_sync() operation is used.

Providing an `enabled=True` argument is already done in pytorch, and is very useful in pytorch in `torch.amp.autocast` and `torch.amp.GradScaler`.

```
if (step % grad_accum_steps + 1) == 0:
    # forward+ backward code
    loss = ddp_model(inputs)
    (loss / grad_accum_steps).backward()
else:
    with ddp_model.no_sync():
         # forward + backward code
         loss = ddp_model(inputs)
        (loss / grad_accum_steps).backward()
```
using the `enabled` argument this can be simplified, preventing bug-prone code duplications:
```
with ddp_model.no_sync(enabled=(step % grad_accum_steps + 1) != 0):
    loss = ddp_model(inputs)
    (loss / grad_accum_steps).backward()
```

The implementation doesn't seem hard, and it will be back-compatible.

### Alternatives

_No response_

### Additional context

DDP with grad accum:
https://discuss.pytorch.org/t/gradient-accumulation-with-ddp-no-sync-interface/169593/3

Current no_sync implementation:
https://github.com/pytorch/pytorch/blob/main/torch/nn/parallel/distributed.py#L1420

torch.amp.autocast enabled=True API:
https://github.com/pytorch/pytorch/blob/09c950cc872dfcee453307db47fa10553c3f5616/torch/amp/autocast_mode.py#L222

### PR DESCRIPTION
**Pull request by filipviz**
Fixes #143721

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -1456,15 +1456,24 @@ def no_sync(self):
...
 
         .. warning::
             The forward pass should be included inside the context manager, or
             else gradients will still be synchronized.
+
+        Args:
+            enabled (bool, optional): Whether this context manager is enabled.
+                ``no_sync`` does nothing when ``enabled=False``. Default: ``True``
         """
+        if not enabled:
```
Comment (wconstab) [inline torch/nn/parallel/distributed.py]: nit: the way you implemented it, i think nested contexts would not work as expected (e.g. enable outside, then disable inside).  I'm not sure if this really matters in practice (does anyone nest?) but might be better to do it the robust way.

**Review by wconstab — APPROVED** [review]
overall LGTM (sorry for missing the notifications before!)

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (filipviz) [reply torch/nn/parallel/distributed.py]: Good point, this was sloppy. Sorry! Fixed this and updated the docstring.
